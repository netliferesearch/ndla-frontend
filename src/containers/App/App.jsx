/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import React from 'react';
import PropTypes from 'prop-types'
import { connect } from 'react-redux';
import Helmet from 'react-helmet';
import cx from 'classnames'
import { PageContainer } from '../../../lib';
import ReactStickyHeader from 'react-sticky-header';
import { MessageShape } from '../../shapes';
import Masthead from '../Masthead';
import Footer from './components/Footer';
import { getLocale } from '../Locale/localeSelectors';
import { getMessages } from '../Messages/messagesSelectors';
import { getSticky } from '../Masthead/mastheadSelectors'
import Alerts from '../Messages/Alerts';
import { injectT } from '../../i18n';

export class App extends React.Component {
  getChildContext() {
    return {
      locale: this.props.locale,
    };
  }

  render() {
    const { dispatch, children, messages, t, params, sticky } = this.props;
    return (
      <PageContainer>
        <Helmet
          title="NDLA"
          meta={[
            { name: 'description', content: t('meta.description') },
          ]}
        />
         <ReactStickyHeader
          // This will be the sticky strip.
          header={
            <div style={{backgroundColor: 'rgba(255,255,255,0.8)'}} className={('Header_root', { sticky })}>
              <Masthead t={t} params={params} />
            </div>
          }
        >
      <Masthead t={t} params={params} />
  </ReactStickyHeader>

        {children}
        <Footer t={t} />
        <Alerts dispatch={dispatch} messages={messages} />
      </PageContainer>
    );
  }
}

App.propTypes = {
  params: PropTypes.shape({
    subjectId: PropTypes.string,
    topicId: PropTypes.string,
  }).isRequired,
  locale: PropTypes.string.isRequired,
  messages: PropTypes.arrayOf(MessageShape).isRequired,
  dispatch: PropTypes.func.isRequired,
};

App.childContextTypes = {
  locale: PropTypes.string,
};

const mapStateToProps = state => ({
  locale: getLocale(state),
  messages: getMessages(state),
  sticky: getSticky(state)
});

export default connect(mapStateToProps)(injectT(App));
