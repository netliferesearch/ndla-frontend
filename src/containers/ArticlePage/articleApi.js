/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import fetch from 'isomorphic-fetch';
import {
  resolveJsonOrRejectWithError,
  apiResourceUrl,
  headerWithAccessToken,
} from '../../util/apiHelpers';

// const converterBaseUrl = 'https://ndla-article-converter.herokuapp.com/article-converter/raw';
const converterBaseUrl = 'http://localhost:3100/article-converter/raw';
const baseUrl = apiResourceUrl('/article-api/v1/articles');

export const fetchArticle = (id, locale, token) =>
  fetch(`${converterBaseUrl}/${locale}/${id}`, {
    headers: headerWithAccessToken(token),
  }).then(resolveJsonOrRejectWithError);

export const fetchArticles = (ids, token) =>
  fetch(`${baseUrl}?ids=${ids.join(',')}`, {
    headers: headerWithAccessToken(token),
  }).then(resolveJsonOrRejectWithError);
