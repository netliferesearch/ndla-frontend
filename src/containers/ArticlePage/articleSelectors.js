/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { createSelector } from 'reselect';
import { getLocale } from '../Locale/localeSelectors';
import { titleI18N, metaDescriptionI18N } from '../../util/i18nFieldFinder';
import formatDate from '../../util/formatDate';

const getArticleFromState = state => state.articles;

export const getArticleById = articleId =>
  createSelector([getArticleFromState], articles => articles[articleId]);

export const getArticle = articleId =>
  createSelector(
    [getArticleById(articleId), getLocale],
    (article, locale) =>
      article
        ? {
            ...article,
            title: titleI18N(article, locale, true),
            metaDescription: metaDescriptionI18N(article, locale, true),
            created: formatDate(article.created, locale),
            updated: formatDate(article.updated, locale),
            requiredLibraries: article.requiredLibraries
              ? article.requiredLibraries.map(lib => {
                  if (lib.url.startsWith('http://')) {
                    return {
                      ...lib,
                      url: lib.url.replace('http://', 'https://'),
                    };
                  }
                  return lib;
                })
              : [],
          }
        : undefined,
  );
