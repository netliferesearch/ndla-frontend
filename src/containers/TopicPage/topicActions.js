/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { createAction } from 'redux-actions';

export const fetchTopics = createAction('FETCH_TOPICS');
export const setTopics = createAction('SET_TOPICS');
export const fetchTopicArticle = createAction('FETCH_TOPIC_ARTICLE');

export const setTopicIntroductions = createAction('SET_TOPIC_INTRODUCTIONS');
