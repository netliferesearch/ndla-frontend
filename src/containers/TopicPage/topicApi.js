/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import {
  resolveJsonOrRejectWithError,
  apiResourceUrl,
  fetchWithAccessToken,
} from '../../util/apiHelpers';

const baseUrl = apiResourceUrl('/taxonomy/v1');

export const fetchTopics = (subjectId, locale) =>
  fetchWithAccessToken(
    `${baseUrl}/subjects/${subjectId}/topics/?recursive=true&language=${locale}`,
  ).then(resolveJsonOrRejectWithError);
