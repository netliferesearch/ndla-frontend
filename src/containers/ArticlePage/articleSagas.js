/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { take, call, put, select } from 'redux-saga-effects';
import { getLocale } from '../Locale/localeSelectors';
import { getArticle } from './articleSelectors';
import * as constants from './articleConstants';
import * as actions from './articleActions';
import * as api from './articleApi';
import { getAccessToken } from '../App/sessionSelectors';

export function* fetchArticle(id) {
  try {
    const locale = yield select(getLocale);
    const token = yield select(getAccessToken);
    const article = yield call(api.fetchArticle, id, locale, token);
    yield put(actions.setArticle(article));
  } catch (error) {
    // TODO: handle error
    console.error(error); //eslint-disable-line
  }
}

export function* watchFetchArticle() {
  while (true) {
    const { payload: id } = yield take(constants.FETCH_ARTICLE);
    const currentArticle = yield select(getArticle(id));
    if (!currentArticle || currentArticle.id !== id) {
      yield call(fetchArticle, id);
    }
  }
}

export default [
  watchFetchArticle,
];
