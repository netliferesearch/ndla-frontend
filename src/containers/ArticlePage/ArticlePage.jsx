/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import React, { PropTypes, Component } from 'react';
import { connect } from 'react-redux';
import Helmet from 'react-helmet';
import { Hero, OneColumn, TopicBreadcrumb } from 'ndla-ui';
import * as actions from './articleActions';
import * as topicActions from '../TopicPage/topicActions';
import { getTopicPath } from '../TopicPage/topicSelectors';
import { getSubjectById } from '../SubjectPage/subjectSelectors';
import * as subjectActions from '../SubjectPage/subjectActions';
import { getArticle } from './articleSelectors';
import { getLocale } from '../Locale/localeSelectors';
import { ArticleShape, SubjectShape, TopicShape } from '../../shapes';
import { toTopic } from '../../routes';
import Article from './components/Article';
import config from '../../config';


const assets = __CLIENT__ ? window.assets : ( // eslint-disable-line no-nested-ternary
  config.isProduction ? require('../../../htdocs/assets/assets') : require('../../../server/developmentAssets') // eslint-disable-line import/no-unresolved
);

class ArticlePage extends Component {
  componentWillMount() {
    const { fetchArticle, fetchTopics, fetchSubjects, params: { articleId, subjectId } } = this.props;
    fetchArticle(articleId);
    if (subjectId) {
      fetchSubjects();
      fetchTopics({ subjectId });
    }
  }

  componentDidMount() {
    if (window.MathJax) {
      window.MathJax.Hub.Queue(['Typeset', window.MathJax.Hub]);
    }
  }

  componentDidUpdate() {
    if (window.MathJax) {
      window.MathJax.Hub.Queue(['Typeset', window.MathJax.Hub]);
    }
  }

  render() {
    const { article, subject, topicPath, locale } = this.props;
    if (!article) {
      return null;
    }
    const scripts = article.requiredLibraries ? article.requiredLibraries.map(lib => ({ src: lib.url, type: lib.mediaType })) : [];
    if (article.content.indexOf('<math') > -1) {
      scripts.push({ async: true, src: `https://cdn.mathjax.org/mathjax/2.7-latest/MathJax.js?config=/assets/${assets['mathjaxConfig.js']}`, type: 'text/javascript' });
    }

    const metaDescription = article.metaDescription ? { name: 'description', content: article.metaDescription } : {};
    return (
      <div>
        <Hero red>
          <OneColumn cssModifier="narrow">
            <div className="c-hero__content">
              { subject ? <TopicBreadcrumb toSubjects={() => '/'} subjectsTitle="" subject={subject} topicPath={topicPath} toTopic={toTopic}>Du er her:</TopicBreadcrumb> : null }
            </div>
          </OneColumn>
        </Hero>
        <Helmet
          title={`NDLA | ${article.title}`}
          meta={[metaDescription]}
          script={scripts}
        />
        <OneColumn cssModifier="narrow">
          <Article article={article} subject={subject} topicPath={topicPath} locale={locale} />
        </OneColumn>
      </div>
    );
  }
}

ArticlePage.propTypes = {
  params: PropTypes.shape({
    articleId: PropTypes.string.isRequired,
    subjectId: PropTypes.string,
    topicId: PropTypes.string,
  }).isRequired,
  article: ArticleShape,
  locale: PropTypes.string.isRequired,
  fetchArticle: PropTypes.func.isRequired,
  fetchSubjects: PropTypes.func.isRequired,
  fetchTopics: PropTypes.func.isRequired,
  subject: SubjectShape,
  topicPath: PropTypes.arrayOf(TopicShape),
};

const mapDispatchToProps = {
  fetchArticle: actions.fetchArticle,
  fetchSubjects: subjectActions.fetchSubjects,
  fetchTopics: topicActions.fetchTopicArticle,
};

const makeMapStateToProps = (_, ownProps) => {
  const { articleId, subjectId, topicId } = ownProps.params;
  const getArticleSelector = getArticle(articleId);
  const getTopicPathSelector = subjectId && topicId ? getTopicPath(subjectId, topicId) : () => undefined;
  const getSubjectByIdSelector = subjectId ? getSubjectById(subjectId) : () => undefined;
  return state => ({
    article: getArticleSelector(state),
    topicPath: getTopicPathSelector(state),
    subject: getSubjectByIdSelector(state),
    locale: getLocale(state),
  });
};


export default connect(makeMapStateToProps, mapDispatchToProps)(ArticlePage);
