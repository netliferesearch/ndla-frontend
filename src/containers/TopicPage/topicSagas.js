/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { all, take, call, put, select } from 'redux-saga/effects';
import { getTopic, hasFetchedTopicsBySubjectId } from './topicSelectors';
import * as actions from './topicActions';
import {
  getArticleIdFromResource,
  isArticleResource,
} from '../Resources/resourceHelpers';
import { fetchArticle } from '../ArticlePage/articleActions';
import * as articleApi from '../ArticlePage/articleApi';
import * as api from './topicApi';
import { getAccessToken } from '../App/sessionSelectors';

export function* fetchTopicIntroductions(topics) {
  try {
    const ids = topics.filter(isArticleResource).map(getArticleIdFromResource);

    if (ids.length === 0) {
      return;
    }
    const token = yield select(getAccessToken);
    const data = yield call(articleApi.fetchArticles, ids, token);
    yield put(
      actions.setTopicIntroductions({
        topics,
        articleIntroductions: data.results,
      }),
    );
  } catch (error) {
    // TODO: handle error
    console.error(error); //eslint-disable-line
  }
}

export function* fetchTopicArticle(subjectId, topicId) {
  const topic = yield select(getTopic(subjectId, topicId));
  const articleId = getArticleIdFromResource(topic);
  if (articleId) {
    yield put(fetchArticle(articleId));
  }
}

export function* fetchTopics(subjectId) {
  try {
    const token = yield select(getAccessToken);
    const topics = yield call(api.fetchTopics, subjectId, token);
    yield put(actions.setTopics({ topics, subjectId }));
    return topics;
  } catch (error) {
    // TODO: handle error
    console.error(error); //eslint-disable-line
    return [];
  }
}

export function* watchFetchTopics() {
  while (true) {
    const { payload: { subjectId, topicId } } = yield take(actions.fetchTopics);
    const hasFetched = yield select(hasFetchedTopicsBySubjectId(subjectId));
    if (!hasFetched) {
      const topics = yield call(fetchTopics, subjectId);
      yield all([
        call(fetchTopicArticle, subjectId, topicId),
        call(fetchTopicIntroductions, topics),
      ]);
    } else {
      yield call(fetchTopicArticle, subjectId, topicId);
    }
  }
}

export default [watchFetchTopics];
