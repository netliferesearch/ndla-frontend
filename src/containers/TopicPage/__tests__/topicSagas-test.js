/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import testSaga from 'redux-saga-test-plan';
import { call } from 'redux-saga-effects';
import * as sagas from '../topicSagas';
import * as api from '../topicApi';
import * as articleApi from '../../ArticlePage/articleApi';
import { fetchArticle } from '../../ArticlePage/articleActions';
import * as actions from '../topicActions';
import { getAccessToken } from '../../App/sessionSelectors';
import { topics } from './mockTopics';

// import { hasFetchedTopicsBySubjectId, getTopic } from '../topicSelectors';

test('topicSagas fetchTopicArticle if articleId is defined', () => {
  const saga = testSaga(sagas.fetchTopicArticle, 1234);
  saga
    .next()
    // .select(getTopic('urn:subject:1', 'urn:topic:1'))
    .next(topics[0])
    .put({ type: `${fetchArticle}`, payload: '1' })

    .next()
    .isDone();
});

test('topicSagas fetchTopics', () => {
  const token = '12345678';
  const saga = testSaga(sagas.fetchTopics, 1234);
  saga
    .next()
    .select(getAccessToken)
    .next(token)
    .call(api.fetchTopics, 1234, token)

    .next([{ id: '123', name: 'Algebra', parent: undefined }])
    .put({ type: actions.setTopics.toString(), payload: { topics: [{ id: '123', name: 'Algebra', parent: undefined }], subjectId: 1234 } })

    .next()
    .isDone();
});

test('topicSagas watchFetchTopics', () => {
  const saga = testSaga(sagas.watchFetchTopics);
  saga
    .next()
    .take(actions.fetchTopics)

    .next({ payload: { subjectId: 'urn:subject:1', topicId: 'urn:topic:1' } })

    .next(false)

    .call(sagas.fetchTopics, 'urn:subject:1')
    .next(topics)

    .parallel([
      call(sagas.fetchTopicArticle, 'urn:subject:1', 'urn:topic:1'),
      call(sagas.fetchTopicIntroductions, topics),
    ])
    .finish()
    .isDone();
});

test('topicSagas watchFetchTopics should not refetch topics', () => {
  const saga = testSaga(sagas.watchFetchTopics);
  saga
    .next()
    .take(actions.fetchTopics)

    .next({ payload: { subjectId: 1234 } })
    // .select(hasFetchedTopicsBySubjectId(1234))

    .next(true)
    .call(sagas.fetchTopicArticle, 1234, undefined)
    .finish()
    .isDone();
});

test('topicSagas fetchTopicIntroductions', () => {
  const mockTopics = [{ contentUri: 'urn:article:1' }, { contentUri: 'urn:learningpath:2' }, { contentUri: 'urn:article:1331' }, { id: 3 }];
  const token = '12345678';
  const saga = testSaga(sagas.fetchTopicIntroductions, mockTopics);
  const data = { results: [{ id: '1', intro: 'Test' }, { id: '1331', intro: 'Test' }] };
  saga
    .next()
    .select(getAccessToken)
    .next(token)
    .call(articleApi.fetchArticles, ['1', '1331'], token)

    .next(data)
    .put({ type: actions.setTopicIntroductions.toString(), payload: { topics: mockTopics, articleIntroductions: data.results } })

    .next()
    .isDone();
});

test('topicSagas fetchTopicIntroductions do not call fetchArticles if no valid ids', () => {
  const mockTopics = [{ contentUri: 'urn:learningpath:2' }];
  const saga = testSaga(sagas.fetchTopicIntroductions, mockTopics);
  saga
    .next()
    .isDone();
});
