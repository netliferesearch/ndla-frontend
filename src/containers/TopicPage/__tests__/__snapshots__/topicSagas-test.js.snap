// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resourceSagas watchFetchTopics 1`] = `
Object {
  "call": Array [
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          "urn:subject:1",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/fetchTopics",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          Object {},
          "@@redux-saga-test-plan/json/function/refineYieldedValue",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/sagaWrapper",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          "urn:subject:1",
          "en",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/<anonymous>",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          "urn:subject:1",
          "urn:topic:1",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/fetchTopicArticle",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          Array [
            Object {
              "contentUri": "urn:article:1",
              "id": "urn:topic:1",
              "name": "Idéutvikling og mediedesign",
              "parent": "urn:subject:1",
            },
            Object {
              "id": "urn:topic:1_1",
              "name": "Mediedesign",
              "parent": "urn:topic:1",
            },
            Object {
              "contentUri": "urn:article:1_2",
              "id": "urn:topic:1_2",
              "name": "Idéutvikling",
              "parent": "urn:topic:1",
            },
            Object {
              "id": "urn:topic:1_2_1",
              "name": "Mediebransjen",
              "parent": "urn:topic:1_2",
            },
            Object {
              "id": "urn:topic:2",
              "name": "Mediekommunikasjon",
            },
          ],
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/fetchTopicIntroductions",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          Object {},
          "@@redux-saga-test-plan/json/function/refineYieldedValue",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/sagaWrapper",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          Object {},
          "@@redux-saga-test-plan/json/function/refineYieldedValue",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/sagaWrapper",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          Array [
            "1",
            "1_2",
          ],
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/<anonymous>",
      },
    },
  ],
  "put": Array [
    Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "payload": Object {
            "subjectId": "urn:subject:1",
            "topics": Array [
              Object {
                "contentUri": "urn:article:1",
                "id": "urn:topic:1",
                "name": "Idéutvikling og mediedesign",
                "parent": "urn:subject:1",
              },
              Object {
                "id": "urn:topic:1_1",
                "name": "Mediedesign",
                "parent": "urn:topic:1",
              },
              Object {
                "contentUri": "urn:article:1_2",
                "id": "urn:topic:1_2",
                "name": "Idéutvikling",
                "parent": "urn:topic:1",
              },
              Object {
                "id": "urn:topic:1_2_1",
                "name": "Mediebransjen",
                "parent": "urn:topic:1_2",
              },
              Object {
                "id": "urn:topic:2",
                "name": "Mediekommunikasjon",
              },
            ],
          },
          "type": "SET_TOPICS",
        },
        "channel": null,
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "PUT": Object {
        "action": Object {
          "payload": Object {
            "articleIntroductions": Array [
              Object {
                "id": "urn:article:1",
                "introduction": "Introduction 1",
              },
              Object {
                "id": "urn:article:1_2",
                "introduction": "Introduction 2",
              },
            ],
            "topics": Array [
              Object {
                "contentUri": "urn:article:1",
                "id": "urn:topic:1",
                "name": "Idéutvikling og mediedesign",
                "parent": "urn:subject:1",
              },
              Object {
                "id": "urn:topic:1_1",
                "name": "Mediedesign",
                "parent": "urn:topic:1",
              },
              Object {
                "contentUri": "urn:article:1_2",
                "id": "urn:topic:1_2",
                "name": "Idéutvikling",
                "parent": "urn:topic:1",
              },
              Object {
                "id": "urn:topic:1_2_1",
                "name": "Mediebransjen",
                "parent": "urn:topic:1_2",
              },
              Object {
                "id": "urn:topic:2",
                "name": "Mediekommunikasjon",
              },
            ],
          },
          "type": "SET_TOPIC_INTRODUCTIONS",
        },
        "channel": null,
      },
    },
  ],
  "select": Array [
    Object {
      "@@redux-saga/IO": true,
      "SELECT": Object {
        "args": Array [],
        "selector": "@@redux-saga-test-plan/json/function/<anonymous>",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "SELECT": Object {
        "args": Array [],
        "selector": "@@redux-saga-test-plan/json/function/<anonymous>",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "SELECT": Object {
        "args": Array [],
        "selector": "@@redux-saga-test-plan/json/function/<anonymous>",
      },
    },
  ],
  "take": Array [
    Object {
      "@@redux-saga/IO": true,
      "TAKE": Object {
        "pattern": "@@redux-saga-test-plan/json/function/actionCreator",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "TAKE": Object {
        "pattern": "@@redux-saga-test-plan/json/function/actionCreator",
      },
    },
  ],
}
`;
